dfkernel/ipkernel.py

```diff
@@ -431,6 +476,73 @@ async def run_cell(*args, **kwargs):
 
         return reply_content, res
 
+    def update_code_cells(self, dfmetadata, update_persistent_code=False):
+        curr_output_tags = dict()
+        updated_code_dict = {}
+        updated_persistent_code_dict = {}
+        code_refs = dict()
+
+        for id, tags in dfmetadata['output_tags'].items():
+            for tag in tags:
+                if isinstance(tag, str):
+                    curr_output_tags.setdefault(tag, set()).add(id)
+        
+        for uuid, refs in dfmetadata['all_refs'].items():
+            for ref_id, ref_tags in refs['ref'].items():
+                for tag in ref_tags:
+                    if code_refs.get(tag):
```

Comment:
Also it appears you have mixed usage of attribute setting as well, being consistent is the most important thing here

Reply:
Used defaultdict  and refactored code to be consistent everywhere